<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Create a standardized VCF file"><title>General function for ingesting GWAS summary statistics into GWAS-VCF format — format_sumstat2vcf • GiTIPs2</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="General function for ingesting GWAS summary statistics into GWAS-VCF format — format_sumstat2vcf"><meta property="og:description" content="Create a standardized VCF file"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">GiTIPs2</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>General function for ingesting GWAS summary statistics into GWAS-VCF format</h1>
      
      <div class="d-none name"><code>format_sumstat2vcf.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Create a standardized VCF file</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">format_sumstat2vcf</span><span class="op">(</span></span>
<span>  <span class="va">path</span>,</span>
<span>  <span class="va">SNP</span>,</span>
<span>  <span class="va">CHR</span>,</span>
<span>  <span class="va">BP</span>,</span>
<span>  <span class="va">A1</span>,</span>
<span>  <span class="va">A2</span>,</span>
<span>  <span class="va">FRQ</span>,</span>
<span>  <span class="va">BETA</span>,</span>
<span>  <span class="va">SE</span>,</span>
<span>  <span class="va">P</span>,</span>
<span>  <span class="va">NCAS</span>,</span>
<span>  <span class="va">N</span>,</span>
<span>  <span class="va">ref_genome</span>,</span>
<span>  <span class="va">id</span>,</span>
<span>  <span class="va">studytype</span>,</span>
<span>  <span class="va">save_path</span>,</span>
<span>  <span class="va">hg19_1kg_phase3</span>,</span>
<span>  <span class="va">trait</span>,</span>
<span>  group_name <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  year <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  author <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  consortium <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  sex <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  pmid <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  population <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  unit <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  subcategory <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  ontology <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  note <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  doi <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  sd <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  <span class="va">nThread</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>path</dt>
<dd><p>the file name with a path which the summary statistics is to be read from.This can be a compressed file as *.tsv.gz</p></dd>


<dt>SNP</dt>
<dd><p>column name for SNP (rs id). If NULL, rs id inferred from reference genome by MungeSumstat</p></dd>


<dt>CHR</dt>
<dd><p>column name for chromosome. Required.</p></dd>


<dt>BP</dt>
<dd><p>column name for base pair positions. Required.</p></dd>


<dt>A1</dt>
<dd><p>column name for reference allele. If NULL, inferred by MungeSumstat.</p></dd>


<dt>A2</dt>
<dd><p>column name for alternative allele. Required.  A note on the allele flipping check: MungeSumstats infers the effect allele will always be the A2 allele,
this is the approach done for IEU GWAS VCF and has such also been adopted here.</p></dd>


<dt>FRQ</dt>
<dd><p>column name for alternate allele frequency of the SNP.If there is no column for FRQ, set as NULL, inferred from 1kg reference</p></dd>


<dt>BETA</dt>
<dd><p>column name for effect size estimate relative to the alternative allele. If NULL, MungeSumstats sets impute_beta = TRUE</p></dd>


<dt>SE</dt>
<dd><p>column name for standard error of effect size estimate. If NULL, then MungeSumstats sets impute_se = TRUE</p></dd>


<dt>P</dt>
<dd><p>column name for unadjusted p-value of effect estimate. Required.</p></dd>


<dt>NCAS</dt>
<dd><p>column name for number of cases, if studytype is "Binary",Required.If there is no column for NCAS, input an integer number. if studytype is "Continuous", NA.</p></dd>


<dt>N</dt>
<dd><p>column name for total sample size. Required.
If there is no column for sample size, then input an integer, MungeSumstats sets compute_n to this integer value the N (sample size) for every SNP in the dataset</p></dd>


<dt>ref_genome</dt>
<dd><p>name of the reference genome used for the GWAS ("GRCh37" or "GRCh38") Argument is case-insensitive. Required.</p></dd>


<dt>id</dt>
<dd><p>name of the vcf output file and used in the gwas catalogue as ID identifier</p></dd>


<dt>studytype</dt>
<dd><p>"Continuous" for quantitative trait and "Binary" for binary trait.</p></dd>


<dt>save_path</dt>
<dd><p>the path which the vcf will be written out. End with "/"</p></dd>


<dt>hg19_1kg_phase3</dt>
<dd><p>if FRQ is NULL, annotate the FRQ with the 1000 genomes phase 3 (hg19/GRCh37) allele frequencies and index
by "/cloud-home/U1006974/GiTIPs/data/reference/ALL.wgs.phase3_shapeit2_mvncall_integrated_v5c.20130502.sites.AF.tsv"</p></dd>


<dt>trait</dt>
<dd><p>trait name. Required.</p></dd>


<dt>group_name</dt>
<dd><p>group name. Optional. "public" or "private" or ...</p></dd>


<dt>year</dt>
<dd><p>Optional.</p></dd>


<dt>author</dt>
<dd><p>Optional.</p></dd>


<dt>consortium</dt>
<dd><p>Optional.</p></dd>


<dt>sex</dt>
<dd><p>Optional.</p></dd>


<dt>pmid</dt>
<dd><p>Optional.</p></dd>


<dt>population</dt>
<dd><p>Optional.</p></dd>


<dt>unit</dt>
<dd><p>Optional.</p></dd>


<dt>subcategory</dt>
<dd><p>Optional.</p></dd>


<dt>ontology</dt>
<dd><p>Optional.</p></dd>


<dt>note</dt>
<dd><p>Optional.</p></dd>


<dt>doi</dt>
<dd><p>Optional.</p></dd>


<dt>sd</dt>
<dd><p>Optional.</p></dd>


<dt>nThread</dt>
<dd><p>Number of threads to use for parallel processes</p></dd>


<dt>...</dt>
<dd><p>parameters from MungeSumstats. See the reference below</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>one .vcf.gz file, one .vcf.gz.tbi file, one a log file .log, and one catalog file, .clg.
Note: The majority of VCF handling bioinformatics libraries use a .vcf.gz suffix, even for block gziped output.
However, VariantAnnotation::writeVcf() even with index = TRUE does not support this and forceably sets the suffix to .bgz.
See https://github.com/Bioconductor/VariantAnnotation/issues/35.
So the current solution is to rename from vcf.bgz to vcf.gz</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p><a href="https://github.com/MRCIEU/gwas-vcf-specification" class="external-link">https://github.com/MRCIEU/gwas-vcf-specification</a></p>
<p><a href="https://neurogenomics.github.io/MungeSumstats/reference/index.html" class="external-link">https://neurogenomics.github.io/MungeSumstats/reference/index.html</a></p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Hao He.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

