<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content='Obtain top hits from a GWAS vcf file with/without clumping
Note: gwasglue::clump_gwasvcf function clumping using 1kg reference genome or other reference genome UKbiobank,
the issue is that if a top variant is missing from the reference genome, it will it will "Removing top variants absence from LD reference panel",
 which is not ideal. As GWAS summary statistic are from imputed genome, for sure it will cover more SNPs than the unimputed reference genome.
The current solution is that get the top SNP in each chromosome (unclumped_top), then do the clumping using reference genome (clumped_df),
if (clumped_df) miss a chromosome in (unclumped_top), combine it into (clumped_df)'><title>Function for obtaining top hits from a GWAS vcf file with and without clumping — tophits_vcf • GiTIPs2</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function for obtaining top hits from a GWAS vcf file with and without clumping — tophits_vcf"><meta property="og:description" content='Obtain top hits from a GWAS vcf file with/without clumping
Note: gwasglue::clump_gwasvcf function clumping using 1kg reference genome or other reference genome UKbiobank,
the issue is that if a top variant is missing from the reference genome, it will it will "Removing top variants absence from LD reference panel",
 which is not ideal. As GWAS summary statistic are from imputed genome, for sure it will cover more SNPs than the unimputed reference genome.
The current solution is that get the top SNP in each chromosome (unclumped_top), then do the clumping using reference genome (clumped_df),
if (clumped_df) miss a chromosome in (unclumped_top), combine it into (clumped_df)'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">GiTIPs2</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Function for obtaining top hits from a GWAS vcf file with and without clumping</h1>
      
      <div class="d-none name"><code>tophits_vcf.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Obtain top hits from a GWAS vcf file with/without clumping
Note: gwasglue::clump_gwasvcf function clumping using 1kg reference genome or other reference genome UKbiobank,
the issue is that if a top variant is missing from the reference genome, it will it will "Removing top variants absence from LD reference panel",
 which is not ideal. As GWAS summary statistic are from imputed genome, for sure it will cover more SNPs than the unimputed reference genome.
The current solution is that get the top SNP in each chromosome (unclumped_top), then do the clumping using reference genome (clumped_df),
if (clumped_df) miss a chromosome in (unclumped_top), combine it into (clumped_df)</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">tophits_vcf</span><span class="op">(</span></span>
<span>  <span class="va">path</span>,</span>
<span>  clump_p <span class="op">=</span> <span class="fl">5e-08</span>,</span>
<span>  clump_r2 <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  clump_kb <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  <span class="va">ldref</span>,</span>
<span>  plink_bin <span class="op">=</span> <span class="fu">genetics.binaRies</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/genetics.binaRies/man/get_plink_binary.html" class="external-link">get_plink_binary</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  <span class="va">nThread</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>path</dt>
<dd><p>Specify the full path starting with /cloud-home/ to the vcf file, including the file name.</p></dd>


<dt>clump_p</dt>
<dd><p>p value threshold to filter. Default = 5e-8</p></dd>


<dt>clump_r2</dt>
<dd><p>Clumping r2 threshold. Default: 0.001</p></dd>


<dt>clump_kb</dt>
<dd><p>Clumping kb window. Default: 10000. (very strict)</p></dd>


<dt>ldref</dt>
<dd><p>LD reference data. Normally Europeans from the 1000 genomes reference panel. Default:"/cloud-home/U1006974/NeuroID/data/1kg/EUR"</p></dd>


<dt>plink_bin</dt>
<dd><p>specify path to plink binary. Default = "genetics.binaRies::get_plink_binary()"</p></dd>


<dt>nThread</dt>
<dd><p>Number of threads to use for parallel processes</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>a data.frame for top hits from a GWAS vcf file with/without clumping</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p><a href="https://mrcieu.github.io/gwasglue/reference/clump_gwasvcf.html" class="external-link">https://mrcieu.github.io/gwasglue/reference/clump_gwasvcf.html</a></p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Hao He.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

